File: leetcode/tree/flatten_binary_tree_linked_list.py
from common import TreeNode

class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        tmp = []
        def solve(root):
            if not root:
                return
            tmp.append(root.val)
            solve(root.left)
            solve(root.right)
        
        solve(root)
        if tmp:
            tmp.pop(0)
            while tmp:
                if root:
                    root.left = None
                    root.right = TreeNode(tmp.pop(0))
                    root = root.right

if __name__ == "__main__":
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.left.left = TreeNode(3)
    root.left.right = TreeNode(4)
    root.right = TreeNode(5)
    root.right.right = TreeNode(6)
    Solution().flatten(root)
    
    while root:
        print(root.val)
        root = root.right

File: leetcode/tree/flatten_binary_tree_linked_list.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	var path []int
	solve(root, &path)
	if path != nil {
		path = path[1:]
		for len(path) != 0 {
			if root != nil {
				root.Left = nil
				root.Right = &TreeNode{path[0], nil, nil}
				root = root.Right
				path = path[1:]
			}
		}
	}
}

func solve(root *TreeNode, path *[]int) {
	if root == nil {
		return
	}
	*path = append(*path, root.Val)
	solve(root.Left, path)
	solve(root.Right, path)
}

func main() {
	root := &TreeNode{1, nil, nil}
	root.Left = &TreeNode{2, nil, nil}
	root.Left.Left = &TreeNode{3, nil, nil}
	root.Left.Right = &TreeNode{4, nil, nil}
	root.Right = &TreeNode{5, nil, nil}
	root.Right.Right = &TreeNode{6, nil, nil}
	flatten(root)
	for root != nil {
		fmt.Println(root.Val)
		root = root.Right
	}
}
